### PR TITLE
Adding a Unit Tests suite

### PR DESCRIPTION
**Pull request by FJBDev**
Since I became a contributor to MyTourbook two years ago (This was my [first commit](https://github.com/wolfgang-ch/mytourbook/commit/4346600b7d70d61906584b7d2142788d7ed54108) on 10/01/2018 😊), and as I've always been implementing unit tests when programming, I was looking for unit tests that can be run in order to ensure that any new feature or code modification would not bring any regressions to the software.
Unfortunately, those tests didn't exist and for the first feature I added in MyTourbook, I decided to [create my own unit tests](https://github.com/wolfgang-ch/mytourbook/commit/3bd688036ce0e2d1679980463d1f18fc5d5208ed).
Thanks to those tests, they allowed me to make sure that no bugs or regressions were introduced before releasing new versions of MyTourbook. For example, I found once a regression [I introduced ](https://github.com/wolfgang-ch/mytourbook/commit/09be84ca9875777dd2a5e97ace4458cecdc84f8e) in the elevation gain/loss computation and I was able to catch this bug before releasing the [20.2 version of MT](https://sourceforge.net/projects/mytourbook/files/MyTourbook/20.2.0/).

However, those tests still are isolated to a single feature and they require some configuration to be performed in order to execute the tests which make them hard to run and therefore are run **less frequently** => **not good!**
Hence, here is my attempt at creating **a unified tests suite** that contains a various list of effective unit tests to be done for MT. Right now, considering the size of MyTourbook, the unit test suite is pretty small but this is a beginning and I hope that with time, the test suite will get bigger.

Here is the list of tests I have developed and gathered:

- File import for the following formats : GPX, Garmin FIT, Garmin TCX ,Suunto Ambit 2 XML & SML, Suunto Spartan/9 JSON

```
“You should be able to run all the unit tests with just one command.”
― Robert C. Martin, Clean Code: A Handbook of Agile Software Craftsmanship 
```

### I like the quote above and below is how you can easily run the tests :

image

### You should then get the below results:

image

Now more than ever, I truly believe that having unit tests is very useful, if not mandatory. If you don't agree, I hope this text will convince you.
**Thoughts? Critics ? Suggestions ?** (@wolfgang-ch, @telemaxx, @rtdog)

As depicted [here](https://dzone.com/articles/java-unit-testing-best-practices-how-to-get-the-mo)

**Why Unit Test?**

Unit testing is a proven technique for ensuring software quality, with plenty of benefits. Here are (more than) a few great reasons to unit test:

1. Unit testing **validates** that each piece of your software not only works properly today, but continues to work in the future, providing a solid foundation for future development.
2. Unit testing **identifies defects at early stages** of the production process, which reduces the costs of fixing them in later stages of the development cycle.
3. Unit-tested code is generally **safer to refactor**, since tests can be re-run quickly to validate that behavior has not changed.
4. Writing unit tests forces developers to consider how well the production code is designed in order to make it **suitable for unit testing**, and makes developers look at their code from a **different perspective**, encouraging them to consider corner cases and error conditions in their implementation.
5. Including unit tests in the **code review process** can reveal how the modified or new code is supposed to work. Plus, reviewers can confirm whether the tests are good ones or not.

### CODE REVIEW

**Comment by wolfgang-ch** [comment]
In 2010 I started to test the UI of MT with this tools https://www.qfs.de/en.html. After one day in creating a test for a dialog I gave up, it was just a lot of work.

To create tests for the tour import makes sense which I've done until now manually. For testing the file import, I'm using all the tour files which have been sent to me in the last 15 years. But these files are private and cannot be used for public Unit tests.

mt-import-test

**Comment by FJBDev** [comment]
> 
> 
> In 2010 I started to test the UI of MT with this tools https://www.qfs.de/en.html. After one day in creating a test for a dialog I gave up, it was just a lot of work.

Indeed, like i mentioned above, tests have to be easy to run, otherwise they don't get run often and if they don't get run often, they are almost useless. Hence why I created a unique suite to run ALL the tests with only one click.
UI tests are a good idea too but I don't know if they can be automated.
> 
> To create tests for the tour import makes sense which I've done until now manually. For testing the file import, I'm using all the tour files which have been sent to me in the last 15 years. But these files are private and cannot be used for public Unit tests.
> 
The files I have put in those unit tests are mine and I am ok for them to be public.

**Comment by wolfgang-ch** [comment]
After searching for Eclipse/RCP UI testing, I found this https://projects.eclipse.org/projects/technology.reddeer tool very promising but there is still a lot of work to create the test cases for all UI components in MT.

**Comment by wolfgang-ch** [comment]
I'm getting these failures when running the test

mt-test-result

[net.tourbook.tests 20201126-143501.zip](https://github.com/wolfgang-ch/mytourbook/files/5603335/net.tourbook.tests.20201126-143501.zip)

File: bundles/net.tourbook.tests/test/utils/Comparison.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+   /**
+    * Compares a test transaction against a control transaction.
+    *
+    * @param testTourData
+    *           The generated test TourData object.
+    * @param controlFileName
+    *           The control's file name.
+    */
+   public static void CompareTourDataAgainstControl(final TourData testTourData,
```
Comment (wolfgang-ch) [inline bundles/net.tourbook.tests/test/utils/Comparison.java:54]: Method names should always start with lowercase, Uppercase is used for classnames

File: bundles/net.tourbook.tests/test/utils/Comparison.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+         encoded = Files.readAllBytes(Paths.get(path));
+         return new String(encoded, encoding);
+      } catch (final IOException e) {
+         e.printStackTrace();
+      }
+
+      return UI.EMPTY_STRING;
+   }
+
+   public static TourData RetrieveImportedTour(final HashMap<Long, TourData> newlyImportedTours) {
```
Comment (wolfgang-ch) [inline bundles/net.tourbook.tests/test/utils/Comparison.java:104]: same naming above

File: bundles/net.tourbook.tests/test/utils/Comparison.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+   }
+
+   /**
+    * Code useful when the tests fail and one wants to be able to compare the expected vs actual
+    * file
+    *
+    * @param controlFileName
+    * @param testJson
+    */
+   private static void WriteErroneousFiles(final String controlFileName, final String testJson) {
```
Comment (wolfgang-ch) [inline bundles/net.tourbook.tests/test/utils/Comparison.java:117]: same naming above

**Review by wolfgang-ch — COMMENTED** [review]
@FJBDev  Naming issues

**Comment by wolfgang-ch** [comment]
> The files I have put in those unit tests are **ALL** mine and I am ok for them to be public.

With your public tours and my traveling 1997 and 2013 in this area (which is displayed in the map) we almost met each other.

mt-usa-pictures

**Comment by FJBDev** [comment]
> 
> 
> I'm getting these failures when running the test
> 
> mt-test-result
> 
> [net.tourbook.tests 20201126-143501.zip](https://github.com/wolfgang-ch/mytourbook/files/5603335/net.tourbook.tests.20201126-143501.zip)

Weird, it works on both my Windows and Linux machine. 😕 
Thanks for the zip file, I will investigate

**Comment by FJBDev** [comment]
> 
> 
> > The files I have put in those unit tests are **ALL** mine and I am ok for them to be public.
> 
> With your public tours and my traveling 1997 and 2013 in this area (which is displayed in the map) we almost met each other.
> 
Not exactly, because if you look at this [specific unit test file](https://github.com/wolfgang-ch/mytourbook/blob/FJBDev/bundles/net.tourbook.tests/test/importdata/suunto2/files/log-F783095113000500-2013-05-18T11_00_38-0.xml), I was not living in Colorado yet 😉

**Comment by FJBDev** [comment]
> 
> 
> After searching for Eclipse/RCP UI testing, I found this https://projects.eclipse.org/projects/technology.reddeer tool very promising but there is still a lot of work to create the test cases for all UI components in MT.

Interesting! Did you try it ? It looks promising as it is actively being developed.

**Comment by FJBDev** [comment]
@wolfgang-ch 
Can you uncomment [those lines](https://github.com/wolfgang-ch/mytourbook/blob/14f5cf5eed70487dd663fe600446bd0fc119572f/bundles/net.tourbook.tests/test/utils/Comparison.java#L80) and send me the *-Erroneous.json files generated by the tests ?
It will show what values your computer is generating that are different from the expected values

**Comment by FJBDev** [comment]
Weird, when I look at the tests log you sent me, I see this :
 <failure>java.lang.AssertionError&#13;
	at importdata.gpx.GPX_SAX_HandlerTest.testParse(GPX_SAX_HandlerTest.java:97)&

Which means that the number of [waypoints](https://github.com/wolfgang-ch/mytourbook/blob/14f5cf5eed70487dd663fe600446bd0fc119572f/bundles/net.tourbook.tests/test/importdata/gpx/GPX_SAX_HandlerTest.java#L97) is not equal to 2

Which number do you get ?

**Comment by FJBDev** [comment]
[Here](https://www.vogella.com/tutorials/EclipseRedDeer/article.html) is a good tutorial for RedDeer, it looks very interesting and I might try to do my first UI test 😃

**Comment by FJBDev** [comment]
And RedDeer also has some [videos](https://www.youtube.com/channel/UCDDzu0AjJ31POHQGsN5jWJw/videos) to get started

**Comment by wolfgang-ch** [comment]
> [Here](https://www.vogella.com/tutorials/EclipseRedDeer/article.html) is a good tutorial for RedDeer, it looks very interesting and I might try to do my first UI test 😃

I've saw this tutorial before I wrote my RedDeer https://github.com/wolfgang-ch/mytourbook/pull/252#issuecomment-734287506 and I also saw one RedDeer video but not the collection you've found.

I also found this testing tool https://www.eclipse.org/rcptt/ which now has few support.

**Comment by wolfgang-ch** [comment]
> 
> 
> > After searching for Eclipse/RCP UI testing, I found this https://projects.eclipse.org/projects/technology.reddeer tool very promising but there is still a lot of work to create the test cases for all UI components in MT.
> 
> Interesting! Did you try it ? It looks promising as it is actively being developed.

No, after my experience with https://www.qfs.de/en.html I've never tried it any more

**Comment by wolfgang-ch** [comment]
> 
> 
> @wolfgang-ch
> I can't reproduce the failures.
> Can you uncomment [those lines](https://github.com/wolfgang-ch/mytourbook/blob/14f5cf5eed70487dd663fe600446bd0fc119572f/bundles/net.tourbook.tests/test/utils/Comparison.java#L80) and send me the json files that fail the tests ?
> It will show what values your computer is generating that are different from the expected values.
> In the past, I have seen issues when using computers with different time zones where some TourData properties had different values

Here are your requested file, it's the one which was modified in this test

[1555291925128_183010004848_post_timeline-1.zip](https://github.com/wolfgang-ch/mytourbook/files/5608064/1555291925128_183010004848_post_timeline-1.zip)

**Comment by wolfgang-ch** [comment]
> 
> 
> Weird, when I look at the tests log you sent me, I see this :
> java.lang.AssertionError
> at importdata.gpx.GPX_SAX_HandlerTest.testParse(GPX_SAX_HandlerTest.java:97)&
> 
> Which means that the number of [waypoints](https://github.com/wolfgang-ch/mytourbook/blob/14f5cf5eed70487dd663fe600446bd0fc119572f/bundles/net.tourbook.tests/test/importdata/gpx/GPX_SAX_HandlerTest.java#L97) is not equal to 2
> 
> Which number do you get ?

In the debug view you can see all needed data, waypoint2 is not imported, I also attached the eclipse log

[mt-log.zip](https://github.com/wolfgang-ch/mytourbook/files/5608069/mt-log.zip)

mt-gpx-assert-2

**Comment by FJBDev** [comment]
I have fixed the unit test issues you have encountered in this [PR](https://github.com/wolfgang-ch/mytourbook/pull/253)

Let me know if you still see failures

File: bundles/net.tourbook.tests/test/utils/Comparison.java (hunk 1 above)
Comment (FJBDev) [reply bundles/net.tourbook.tests/test/utils/Comparison.java:54]: Fixed in my [PR](https://github.com/wolfgang-ch/mytourbook/pull/255)

File: bundles/net.tourbook.tests/test/utils/Comparison.java (hunk 2 above)
Comment (FJBDev) [reply bundles/net.tourbook.tests/test/utils/Comparison.java:104]: Fixed in my [PR](https://github.com/wolfgang-ch/mytourbook/pull/255)

File: bundles/net.tourbook.tests/test/utils/Comparison.java (hunk 3 above)
Comment (FJBDev) [reply bundles/net.tourbook.tests/test/utils/Comparison.java:117]: Fixed in my [PR](https://github.com/wolfgang-ch/mytourbook/pull/255)

**Comment by FJBDev** [comment]
> > [Here](https://www.vogella.com/tutorials/EclipseRedDeer/article.html) is a good tutorial for RedDeer, it looks very interesting and I might try to do my first UI test smiley
> 
> I've seen this tutorial before I wrote my RedDeer [#252 (comment)](https://github.com/wolfgang-ch/mytourbook/pull/252#issuecomment-734287506) and I also saw one RedDeer video but not the collection you've found.
> 
> I also found this testing tool https://www.eclipse.org/rcptt/ which now has few support.

Can you give me examples of what UI tests you think should useful to be done ?

**Comment by wolfgang-ch** [comment]
I see the most problematic part in the UI are the different selection/event providers/listeners.

In the beginning of MT only the part selection provider/listeners were used but parts are firing events only when the part is active. So when you hover over the day statistic and scroll the mouse wheel, it selects the subsequent tours but when this part is not active, the newly selected tours would not be recognized by other parts, e.g. map or tour chart. For this reason the tour events were introduced -> this is the reason why there is now a mix of different event/selection listener/providers.

Because some event listeners do also fire events, it gets more and more complicated. For me this would be a good example to do a UI unit test with different opened views which are event/selection listeners/providers. I'm aware that this would not be an easy unit test but it would be one which could help to find event/selection issues.

I'm not sure if a UI unit test of a normal dialog would be helpful however some slideouts are also fireing events, e.g. tour filter which could be also a candidate for a UI unit test.

**Comment by FJBDev** [comment]
Ok, good info.

I have looked into RedDeer and RCPTT and so far RedDeer seems a bit better.

My goal now is to understand how to create a test that can launch MT with a dummy database.
